File: GGAndroid_3.0/core/support/rxjava/src/main/java/io/ganguo/rx/RxActivityResult.java
package io.ganguo.rx;

import android.app.Activity;
import android.content.Intent;

import androidx.fragment.app.Fragment;
import io.reactivex.Observable;

/**
 * Created by Roger on 8/10/16.
 */
public class RxActivityResult {
    /**
     * StartActivity in rx way
     *
     * @param activity    a activity implement {@link ActivityResultOwner}
     * @param intent      The intent to start
     * @param requestCode If >= 0, this code will be returned in
     *                    {@link ActivityResult} when the activity exits.
     * @return a Observable emit {@link ActivityResult}
     */
    @SuppressWarnings("unchecked raw type")
    public static Observable<ActivityResult> startIntent(Activity activity, Intent intent, int requestCode) {
        if (activity instanceof ActivityResultOwner) {
            activity.startActivityForResult(intent, requestCode);
            return ((ActivityResultOwner) activity).getActivityResult().hide();
        } else {
            return null;
        }
    }

    @SuppressWarnings("unchecked raw type")
    public static Observable<ActivityResult> startIntent(Activity activity, Intent intent) {
        if (activity instanceof ActivityResultOwner) {
            activity.startActivityForResult(intent, 0);
            return ((ActivityResultOwner) activity).getActivityResult().hide();
        } else {
            return null;
        }
    }

    /**
     * StartActivity in rx way
     *
     * @param fragment    a fragment implement {@link ActivityResultOwner}
     * @param intent      The intent to start
     * @param requestCode If >= 0, this code will be returned in
     *                    {@link ActivityResult} when the activity exits.
     * @return a Observable emit {@link ActivityResult}
     */
    @SuppressWarnings("unchecked raw type")
    public static Observable<ActivityResult> startIntent(Fragment fragment, Intent intent, int requestCode) {
        if (fragment instanceof ActivityResultOwner) {
            fragment.startActivityForResult(intent, requestCode);
            return ((ActivityResultOwner) fragment).getActivityResult().hide();
        } else {
            return null;
        }
    }

    @SuppressWarnings("unchecked raw type")
    public static Observable<ActivityResult> startIntent(Fragment fragment, Intent intent) {
        if (fragment instanceof ActivityResultOwner) {
            fragment.startActivityForResult(intent, 0);
            return ((ActivityResultOwner) fragment).getActivityResult().hide();
        } else {
            return null;
        }
    }
}

File: GGAndroid_4.0/core/support/support-rx-activity-result/src/main/java/io/ganguo/rxresult/RxActivityResult.kt
package io.ganguo.rxresult

import android.app.Activity
import android.content.Intent
import androidx.fragment.app.Fragment
import io.reactivex.rxjava3.core.Observable

/**
 * <pre>
 *   @author : leo
 *   @time   : 2020/10/14
 *   @desc   : Activity Result based on RxJava implementation
 * </pre>
 */
class RxActivityResult {

    companion object {
        /**
         * StartActivity in rx way
         * @param activity Activity
         * @param intent Intent?
         * @param requestCode Int
         * @return Observable<ActivityResult?>?
         */
        @JvmStatic
        fun startIntent(activity: Activity, intent: Intent?, requestCode: Int = 0, takeCount: ActivityResultOwner.TakeCount = ActivityResultOwner.TakeCount.ONCE): Observable<ActivityResult> = let {
            check(activity is ActivityResultOwner) {
                "Activity must be implementation ActivityResultOwner!!"
            }
            activity.startActivityForResult(intent, requestCode)
            activity.toObservable(takeCount)
        }


        /**
         * StartActivity in rx way
         *
         * @param fragment    a fragment implement [ActivityResultOwner]
         * @param intent      The intent to start
         * @param requestCode If >= 0, this code will be returned in
         * [ActivityResult] when the activity exits.
         * @return a Observable emit [ActivityResult]
         */
        @JvmStatic
        fun startIntent(fragment: Fragment, intent: Intent?, requestCode: Int = 0, takeCount: ActivityResultOwner.TakeCount = ActivityResultOwner.TakeCount.ONCE): Observable<ActivityResult> = let {
            check(fragment is ActivityResultOwner) {
                "Activity must be implementation ActivityResultOwner!!"
            }
            fragment.startActivityForResult(intent, requestCode)
            fragment.toObservable(takeCount)
        }

    }
}


